### ISSUE TITLE
is there a way to use (something like) a Poly on a sealed trait?

### ISSUE THREAD

**Issue by fommil** [milessabin/shapeless#361]
I'd like to be able to remove boilerplate like this:

```scala
sealed trait Trait
case class Foo() extends Trait
case class Bar() extends Trait

sealed trait Thing[T] { def thingy: String}
implicit def foo: Thing[Foo] = ???
implicit def bar: Thing[Bar] = ???

// this is the bit I'd like to simplify...
def thing(t: Trait): String = t match {
  case Foo() => implicitly[Thing[Foo]].thingy
  case Bar() => implicitly[Thing[Bar]].thingy
}
```

If I know which implementation of `Trait` I have, then obviously there is no need for the `thing` method, but it gets really really long if I have quite a few implementations.

If all I'm doing is a side-effect, or a transformation that doesn't change the type signature, then I've found that `everywhere` works quite well (that's probably a sledgehammer to crack a nut). But if I want to change the type (as in this example, where the application of `Thing` converts a `Trait` into `String`) then I have no options that I know of.

Basically, does shapeless have something that I could use here or is this a feature request?

**Comment by fommil** [comment]
btw, is it perhaps better to ask these sorts of questions on stackoverflow? I'm used to using the issue tracker as a Q&A forum but since shapeless has an active community, there might be a better place to ask this sort of thing.

**Comment by fommil** [comment]
am I just being an idiot, is this just basic application of a poly?

**Comment by fommil** [comment]
Almost, I have this:

```scala
  object thing extends Poly1 {
    implicit def thingAt[T <: Trait](implicit t: Thing[T]) = at[T](a => t.thingy)
  }
```

but I can't see how to actually apply this to an instance of a `Trait` and get a `String` back (well, for the compiler to know that).

**Comment by fommil** [comment]
I'll discuss on gitter and close here. Full code at https://gist.github.com/fommil/d3fe97c86deb74a94fe2

**Comment by fommil** [comment]
full example https://gist.githubusercontent.com/fommil/d3fe97c86deb74a94fe2/raw/1c13543baf34a01f949b5f534cc3434ef3565c05/gistfile1.scala

**Comment by fommil** [comment]
reopening because actually this needs me to know the type when passing it to `thing`... I need it to resolve which element in the coproduct to use.

**Comment by milessabin** [comment]
This isn't a bug! You need to map your poly over the coproduct representation of `Trait`.

**Comment by fommil** [comment]
I updated the gist, I don't know how to do the `select` call to get the result back again.

**Comment by milessabin** [comment]
Seriously, this is a StackOverflow question, or something to ask about on the gitter channel. Also check your twitters for a solution ;-)

**Comment by fommil** [comment]
cool, `unify` works! Thanks :smile:

The syntax is a bit messy though, is there any way to do it all within a single `object`?

I can't use gitter at work and stackoverflow is very flaky (random pages banned).

**Comment by fommil** [comment]
if I put that logic in an `apply` on the `object`, it also works, which is cleaner. I'll look into abstracting it to avoid having to write the `Generic / mapping` etc.

Is there a big overhead to obtain the `Generic[T]` or should that be cached?

**Comment by milessabin** [comment]
Gotcha, but please don't use issues for this sort of thing. Send me a mail or something.

**Comment by fommil** [comment]
ok, that really limits me though because I don't get a reliable mobile signal at work :sob: I'll send them at lunchtime or wait till the evenings and use stack overflow.

**Comment by milessabin** [comment]
Use a gist then?

**Comment by fommil** [comment]
Will do! Might write this up tonight as a Q&A on shark overflow

**Comment by fommil** [comment]
http://stackoverflow.com/questions/29651482
